Mathlib/Tactic/Peel.lean

```diff
@@ -0,0 +1,209 @@
+```
+(which differ only in `<`/`≤`), applying `peel h with h_peel ε hε N n hn` will yield a tactic state:
+```
+h : ∀ ε > (0 : ℝ), ∃ N : ℕ, ∀ n ≥ N, 1 / (n + 1 : ℝ) < ε
+ε : ℝ
+hε : 0 < ε
+N n : ℕ
+hn : N ≤ n
+h_peel : 1 / (n + 1 : ℝ) < ε
+⊢ 1 / (n + 1 : ℝ) ≤ ε
+```
+and the goal can be closed with `exact h_peel.le`.
+Note that in this example, `h` and the goal are logically equivalent statements, but `peel`
+*cannot* be immediately applied to show that the goal implies `h`.
+
+`peel` can take an optional numeric argument prior to the supplied hypothesis; in which case it will
+autoname the introduced variables, but they may be inaccessible. However, in this case the user must
+still introduce a single name such as `with h_peel` for the new hypothesis.
+
+In addition, `peel` supports goals of the form `(∀ x, p x) ↔ ∀ x, q x`, or likewise for any
+other quantifier. In this case, there is no hypothesis or term to supply, but otherwise the syntax
+is the same. So for such goals, the syntax is `peel 1` or `peel with x`, and after which the
+resulting goal is `p x ↔ q x`. The `congr!` tactic can also be applied to goals of this form using `congr! 1 with x`. While `congr!` applies congruence lemmas in general, `peel` can be relied upon to only apply to outermost quantifiers.
+
+Finally, the user may supply a term `e` via `... using e` in order to close the goal
+immediately. In particular, `peel h using e` is equivalent to `peel h; exact e`. The `using` syntax
+may be paired with any of the other features of `peel`.
+
+This tactic works by repeatedly applying `forall_imp`, `Exists.imp`, `Filter.Eventually.mp`,
+`Filter.Frequently.mp`, and `Filter.eventually_of_forall` and introducing the variables as these
+are applied.
+-/
+syntax (name := peel) "peel" (num)? (ppSpace colGt term)? (withArgs)? (usingArg)? : tactic
+
+private lemma and_imp_left_of_imp_imp {p q r : Prop} (h : r → p → q) : r ∧ p → r ∧ q := by tauto
+
+/-- This is the core to the `peel` tacitc.
+
+It tries to match `e` and `goal` as quantified statements (using `∀`, `∃`, `∀ᶠ` or `∃ᶠ`), then
+applies `forall_imp`, `Exists.imp`, `Filter.Eventually.mp`, `Filter.Frequently.mp` (the latter two
+also use `Filter.eventually_of_forall`) as appropriate, and then intros two variables, with the
+optionally provided names. If, for example, `goal : ∃ y : α, q y`, the metavariable returned has
+type `q x` where `x : α` has been introduced into the context.
+
+The special casing for `e`/`goal` pairs of type `r ∧ p` and `r ∧ q` exists primarily to deal with
+quantified statements like `∃ δ > (0 : ℝ), q δ`. -/
+def peelQuantifier (goal : MVarId) (e : Expr) (n : Option Name := none) (n' : Option Name := none) :
+    MetaM (Option FVarId × List MVarId) := goal.withContext do
+  let ty ← whnfR (← inferType e)
+  let target ← whnfR (← goal.getType)
+  let n' ← n'.getDM (mkFreshUserName `h_peel)
+  unless (← isProp ty) && (← isProp target) do
+    throwError m!"either {ty} or {target} is not a proposition"
```

Comment:
```suggestion
  guard ((← isProp ty) && (← isProp target)) <|>
    throwError m!"either {ty} or {target} is not a proposition"
```
Does this also work?  Does it speed up the compilation, using it systematically for all the `unless` that you use?

Just trying out equivalent changes...

Reply:
Great, thanks for checking and good to know!